scene/resources/paragraph_layout.cpp

```diff
@@ -0,0 +1,1354 @@
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT startED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "paragraph_layout.h"
+#include "core/os/os.h"
+#include "font.h"
+
+/*************************************************************************/
+/*  ScriptRunBuilder                                                     */
+/*************************************************************************/
+
+bool ParagraphLayout::ScriptRunBuilder::same_script(int32_t p_script_one, int32_t p_script_two) {
+
+	return p_script_one <= USCRIPT_INHERITED || p_script_two <= USCRIPT_INHERITED || p_script_one == p_script_two;
+}
+
+bool ParagraphLayout::ScriptRunBuilder::next() {
+
+	int32_t start_sp = paren_sp; // used to find the first new open character
+	UErrorCode error = U_ZERO_ERROR;
+
+	// if we've fallen off the end of the text, we're done
+	if (script_end >= char_limit) {
+		return false;
+	}
+
+	script_code = USCRIPT_COMMON;
+
+	for (script_start = script_end; script_end < char_limit; script_end += 1) {
+		UChar high = char_array[script_end];
+		UChar32 ch = high;
+
+		// if the character is a high surrogate and it's not the last one
+		// in the text, see if it's followed by a low surrogate
+		if (high >= 0xD800 && high <= 0xDBFF && script_end < char_limit - 1) {
+			UChar low = char_array[script_end + 1];
+
+			// if it is followed by a low surrogate,
+			// consume it and form the full character
+			if (low >= 0xDC00 && low <= 0xDFFF) {
+				ch = (high - 0xD800) * 0x0400 + low - 0xDC00 + 0x10000;
+				script_end += 1;
+			}
+		}
+
+		UScriptCode sc = uscript_getScript(ch, &error);
+		int32_t pair_index = u_getBidiPairedBracket(ch);
+
+		// Paired character handling:
+		//
+		// if it's an open character, push it onto the stack.
+		// if it's a close character, find the matching open on the
+		// stack, and use that script code. Any non-matching open
+		// characters above it on the stack will be poped.
+		if (pair_index >= 0) {
+			if ((pair_index & 1) == 0) {
```

Comment:
Same here; to know if this is an open or close character, you need to check if `UCHAR_BIDI_PAIRED_BRACKET_TYPE` is `U_BPT_OPEN` or `U_BPT_CLOSE`.